proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,326 @@
+    val dataStream: Observable<String>
+        get() = _dataStream
+}
+```
+
+Turns into:
+
+```kotlin
+class MyClass {
+    val dataStream: Observable<String>
+        field = PublishSubject.create()
+}
+```
+
+## Design
+
+The proposed design consists of two new ideas.
+
+### Explicit Backing Fields
+
+```kotlin
+val it: P
+    [visibility] field[: F] = initializer
+```
+
+The above `field` declaration is referred to as an _explicit backing field declaration_.
+
+Explicit backing fields is a FIR-only feature.
+
+#### Accessors
+
+- if `P :> F`, the compiler can provide a default getter
+- if `P <: F`, the compiler can provide a default setter
+
+If the compiler can not provide a getter, the user must declare it explicitly. The same applies to setters in case of `var` properties.
+
+```kotlin
+public val flow: SharedFlow<String>
+    field: MutableSharedFlow? = null
+    get() {
+        return field ?: run { ... }
+    }
+```
+
+#### Visibility
+
+Only the `private` and the `internal` visibilities are allowed for explicit backing fields now. The default visibility is `private`.
+
+```kotlin
+val mutableWithinModule: List<Item>
+    internal field = mutableListOf()
+```
+
+Right now, there is no special syntax to explicitly access the backing field from outside the property accessors, but we will see how it can be accessed implicitly via the ["second idea"](#smart-type-narrowing).
+
+#### Lateinit
+
+If a property has an explicit backing field declaration, and it needs to be `lateinit`, the modifier must be placed at the `field` declaration.
+
+```kotlin
+var someStrangeExample: Int
+    lateinit field: String
+    get() = field.length
+    set(value) {
+        field = value.toString()
+    }
+```
+
+#### Restrictions
+
+If there is an explicit backing field, the property must not declare an initializer. If the explicit backing field is not `lateinit`, it must have an initializer. For `lateinit` properties, initializers are forbidden.
+
+For now, we assume `val` properties have immutable backing fields, and `var` properties have mutable ones. That is, assignment to `field` inside a `val` property getter results in an error.
+
+Explicit backing fields are not allowed inside interfaces or abstract properties, as well as they are forbidden for delegated properties.
+
+### Smart Type Narrowing
+
+#### Rules
+
+If:
+
+1. the compiler can guarantee the property getter returns the same instance as the one stored in the backing field
+2. the type of that instance is compatible with the property type
+3. the backing field visibility allows a hypothetical direct access
+4. there is no ambiguity in such an access
+
+it can then narrow down the returned instance type at the call site without loss of functionality. This is what is meant by the words _smart type narrowing_.
+
+The formal checks corresponding to the above rules are:
+
+1. the property does not have a custom getter, only the default one
+2. `A :> B`
```

Comment:
Please replace with `P :> F` (to be honest, I suggested this rename mainly because of this place)